resolwe/flow/models/utils.py

```diff
@@ -339,45 +339,74 @@ def hydrate_input_uploads(input_, input_schema, hydrate_values=True):
                 value['file_temp'] = 'Invalid value for file_temp in DB'
 
 
-def hydrate_size(data):
+def hydrate_size(data, force=False):
     """Add file and dir sizes.
 
     Add sizes to ``basic:file:``, ``list:basic:file``, ``basic:dir:``
     and ``list:basic:dir:`` fields.
 
+    Force parameter is used when migrating data: one wants to recompute
+    file sizes also on objects that already have these values.
     """
     from .data import Data  # prevent circular import
 
+    def get_dir_size(path):
+        """Get directory size."""
+        total_size = 0
+        for dirpath, _, filenames in os.walk(path):
+            for file_name in filenames:
+                file_path = os.path.join(dirpath, file_name)
+                if not os.path.isfile(file_path):  # Skip (broken) links
+                    continue
+                total_size += os.path.getsize(file_path)
+        return total_size
+
+    def get_refs_size(obj, obj_path):
+        """Caculate size of all references."""
+        total_size = 0
+        for ref in obj.get('refs', []):
+            ref_path = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(data.pk), ref)
+            if os.path.isfile(ref_path):
+                total_size += os.path.getsize(ref_path)
+            elif os.path.isdir(ref_path):
+                total_size += get_dir_size(ref_path)
+            elif os.path.exists(ref_path):
```

Comment:
Move this and the `else` to the validation function: https://github.com/genialis/resolwe/blob/master/resolwe/flow/models/utils.py#L105-L129